benchexec/resources.py

```diff
@@ -57,149 +63,332 @@ def get_cpu_cores_per_run(
-        sys.exit(f"Could not read CPU information from kernel: {e}")
-    return _get_cpu_cores_per_run0(
-        coreLimit,
-        num_of_threads,
-        use_hyperthreading,
-        allCpus,
-        cores_of_unit,
-        siblings_of_core,
-    )
+            for resultlist in prelim_result:
+                result.append(resultlist[:coreLimit])
+        else:
+            i = coreLimit
+            while i >= coreRequirement:
+                if check_distribution_feasibility(
+                    i,
+                    num_of_threads,
+                    use_hyperthreading,
+                    allCpus,
+                    siblings_of_core,
+                    hierarchy_levels,
+                    isTest=True,
+                ):
+                    result = get_cpu_distribution(
+                        i,
+                        num_of_threads,
+                        use_hyperthreading,
+                        allCpus,
+                        siblings_of_core,
+                        hierarchy_levels,
+                    )
+                    break
+                else:
+                    i -= 1
 
+    return result
 
-def _get_cpu_cores_per_run0(
+
+# define class VirtualCore to generate core objects
+class VirtualCore:
+    """
+    Generates an object for each available CPU core,
+    providing its ID and a list of the memory regions it belongs to.
+    @attr coreId: int returned from the system to identify a specific core
+    @attr memory_regions: list with the ID of the corresponding regions the core belongs too sorted according to its size
+    """
+
+    def __init__(self, coreId, memory_regions=None):
+        self.coreId = coreId
+        self.memory_regions = memory_regions
+
+    def __str__(self):
+        return str(self.coreId) + " " + str(self.memory_regions)
+
+
+def filter_hyperthreading_siblings(allCpus, siblings_of_core, hierarchy_levels):
+    """
+    Deletes all but one hyperthreading sibling per physical core out of allCpus, siblings_of_core & hierarchy_levels
+    @param allCpus:         list of VirtualCore objects
+    @param siblings_of_core:mapping from one of the sibling cores to the list of siblings including the core itself
+    """
+    for core in siblings_of_core:
+        no_HT_filter = []
+        for sibling in siblings_of_core[core]:
+            if sibling != core:
+                no_HT_filter.append(sibling)
+        for virtual_core in no_HT_filter:
+            siblings_of_core[core].remove(virtual_core)
+            region_keys = allCpus[virtual_core].memory_regions
+            i = 1
+            while i < len(region_keys):
+                hierarchy_levels[i][region_keys[i]].remove(virtual_core)
+                i = i + 1
+            allCpus.pop(virtual_core)
+
+
+def check_distribution_feasibility(
     coreLimit,
     num_of_threads,
     use_hyperthreading,
     allCpus,
-    cores_of_unit,
     siblings_of_core,
+    hierarchy_levels,
+    isTest=True,
```

Comment:
The less production code is affected by specific needs of the test, the better. It is possible to do this if it is necessary, though. But here I don't really see a need: Couldn't the test simply check that the exception is thrown instead?

Reply:
Unfortunately, I don't understand your feedback here.
I wrote a "check feasibility" function that is used in 2 places:

1. to determine the highest possible coreLimit 
```
        i = coreLimit
            while i >= coreRequirement:
                if check_distribution_feasibility(
                    i,
                    num_of_threads,
                    allCpus,
                    siblings_of_core,
                    hierarchy_levels,
                    isTest=True,
                ):
                    break
                else:
                    i -= 1
```

2 . in the assignment, leading to a sys.exit if the input parameters dont add up

```
    check_distribution_feasibility(
        coreLimit,
        num_of_threads,
        allCpus,
        siblings_of_core,
        hierarchy_levels,
        isTest=False,
    )
```

Reply:
Ah, I think now I understand. `isTest` is not related to running this in unit tests?

I guess it would still be cleaner to always have it throw an exception and catch it in the caller instead of this parameter. The alternative would be to make it return `Optional[str]` with the error message and throw if the result is not None, but I don't think this is better.